libsplinter/src/network/connection_manager/mod.rs

```diff
@@ -0,0 +1,687 @@
+        let (sender, recv) = unbounded();
+        let connection_state = self.connection_state.clone();
+
+        let join_handle = thread::Builder::new()
+            .name("Connection Manager".into())
+            .spawn(move || {
+                let mut subscribers = HashMap::new();
+                loop {
+                    match recv.recv() {
+                        Ok(CmMessage::Shutdown) => break,
+                        Ok(CmMessage::Subscribe(id, sender)) => {
+                            subscribers.insert(id, sender);
+                        }
+                        Ok(CmMessage::UnSubscribe(ref id)) => {
+                            subscribers.remove(id);
+                        }
+                        Ok(CmMessage::Request(req)) => {
+                            let mut state = match connection_state.lock() {
+                                Ok(s) => s,
+                                Err(err) => {
+                                    notify_subscribers(
+                                        &mut subscribers,
+                                        vec![CmNotification::FatalError {
+                                            error: ConnectionManagerError::StatePoisoned,
+                                            message: format!("State has been poisoned {:?}", err),
+                                        }],
+                                    );
+                                    break;
+                                }
+                            };
+
+                            handle_request(req, &mut state);
+                        }
+                        Ok(CmMessage::HeartbeatNotifications(notifications)) => {
+                            notify_subscribers(&mut subscribers, notifications)
+                        }
+                        Err(_) => {
+                            warn!("All senders have disconnected");
+                            break;
+                        }
+                    }
+                }
+            })?;
+
+        self.hb_monitor.start(sender.clone())?;
+        self.join_handle = Some(join_handle);
+        self.shutdown_handle = Some(ShutdownHandle {
+            sender: sender.clone(),
+            hb_shutdown_handle: self.hb_monitor.shutdown_handle().unwrap(),
+        });
+        self.sender = Some(sender.clone());
+
+        Ok(Connector { sender })
+    }
+
+    pub fn shutdown_handle(&self) -> Option<ShutdownHandle> {
+        self.shutdown_handle.clone()
+    }
+
+    pub fn await_shutdown(self) {
+        self.hb_monitor.await_shutdown();
+
+        let join_handle = if let Some(jh) = self.join_handle {
+            jh
+        } else {
+            return;
+        };
+
+        if let Err(err) = join_handle.join() {
+            error!(
+                "Connection manager thread did not shutdown correctly: {:?}",
+                err
+            );
+        }
+    }
+
+    pub fn shutdown_and_wait(self) {
```

Comment:
Not sure this should be public, as I don't think the internals of this are useful.  It seems like a test-only convenience.